settings: Add setting for the first channel number used

herrnst opened the pull request:
Make the first channel number used to populate the channel list configurable as the default of 1 very likely conflicts with other PVR backend channel list when multiple backends are activated.

bkuhls commented:
Works fine and is a really useful patch, thanks!

herrnst commented:
Small, subtle nudge towards @ksooo since this change became three months old a few days ago 🤫

ksooo commented:
If you are running multiple PVR add-ons and want to avoid conflicts re. channel numbers, just give the add-ons different priorities. No need for special support for this in the add-ons.

Settings -> PVR & Live TV -> General -> Client Priorities

bkuhls commented:
Yes, I used this feature with tvheadend and plutotv pvr addons until I found this patch.
The kodi function, afaics, renumbers all stations (yes, tvheadend channels are not mixed with plutotv ones) which is bad for my use case.
I configured blocks of stations in tvheadend, like 1-99, 100-199, 200-299... where in each block there are lots of free numbers making it easier to insert new stations. Yes, I do not like channel groups ;) These numbers from tvheadend get lost with Kodi's client priorities.

herrnst commented:
Exactly what @bkuhls wrote is my usecase aswell, ie. have backend-controlled channel numbers (I don't feel like sorting things myself and I'm used to what my cable operator provides) and that will mix up in a multiple PVR addon setup. Originally I found out about this in the https://github.com/nirvana-7777/pvr.magenta addon (Telekom doesn't do continuous channel numbers aswell and their grouping works fine) where a "first channel number" is configurable, too and thus integrates perfectly in such scenarios, so I made this little modification to pvr.pluto aswell which makes it integrate perfectly aswell. Also, I believe this is a handy feature for others too, so that's why I came up with this PR.

NB this is mostly the same as if one would configure channel numbers in TVheadend (or VDR or anything else) not starting with 1, with "use backend channel numbers" enabled.

ksooo commented on pvr.plutotv/resources/language/resource.language.de_de/strings.po:
```diff
@@ -30,6 +30,10 @@ msgctxt "#30007"
 msgid "(Re)install Widevine CDM library"
 msgstr "Widevine CDM-Bibliothek (erneut) installieren"
 
+msgctxt "#30008"
+msgid "First channel number"
+msgstr "Erste Sendernummer"
+
```
Please remove this from this PR, as translations are handled via Weblate, not GitHub. Only "source" language, which is en_gb is handled via GitHub.

ksooo commented on src/PlutotvData.cpp:
```diff
@@ -262,6 +263,16 @@ std::string PlutotvData::GetSettingsUUID(const std::string& setting)
+{
+  int startnum = 1;
+
+  if(!kodi::addon::CheckSettingInt("start_channelnum", startnum))
+    kodi::Log(ADDON_LOG_DEBUG, "Start channel number not set, defaulting to 1");
```
The `if` and the log line can be removed, as not having the number set is something that should never happens as the setting definition sets a default value and forbids "empty" values.

ksooo commented on src/PlutotvData.h:
```diff
@@ -74,6 +74,7 @@ class ATTR_DLL_LOCAL PlutotvData : public kodi::addon::CAddonBase,
 
   std::string GetChannelStreamURL(int uniqueId);
   std::string GetSettingsUUID(const std::string& setting);
+  int GetSettingsStartChannel();
```
This method most probably can be `const`. => ` int GetSettingsStartChannel() const;`

ksooo reviewed (changes requested):
Some things that need to be changed.

ksooo commented on src/PlutotvData.cpp:
```diff
@@ -113,7 +113,8 @@ bool PlutotvData::LoadChannelsData()
   kodi::Log(ADDON_LOG_DEBUG, "[channels] size: %i;", channelsDoc["result"].Size());
 
-  int i = 0;
+  // Use configured start channel number minus one due to the loop logic
+  int i = (GetSettingsStartChannel() - 1);
```
Instead of adapting to the loop logic, you could cleanup the loop logic.

```c++
int i = GetSettingsStartChannel();
for (const auto& channel : channelsDoc["result"].GetArray())
{
  ...
  PlutotvChannel plutotv_channel;
  plutotv_channel.iChannelNumber = i++; // position
  ...
}
```

herrnst commented:
Thanks for reviewing! Changed as requested and rebased ontop of current Omega HEAD. Note I opted to go for GetSettingInt() instead of CheckSettingInt as this should even provide a default startchannelnum even if the unlikely case happens that no value is returned despite the default in the addon settings definition.

ksooo reviewed (approved):
Looks good. Thanks for your contribution.
